### PR TITLE
add hgraph test & fix

### LINKED ISSUE #40: Add a new Layer "DataCell" to manage data access
I think high level index should ignore the specific implement of quantization and io type, So the quantization and io components can be managed by new layer DataCell which provide different data access interface to the index layer. just like the following show

|Layer| Description|Implement|
|-------|--------|---------|
|Index Layer | ANN algorithm & architecture.          |                       Graph (HNSW,NNDescent) + Reorder,  IVF(PQ,SQ,fastscan), HNSW_IVF, MultiIndex ......|
|DataCell Layer| data access, use quantization, io and others    | FlattenDC, GraphNeighborhoodDC, MixtureDC .......|
|Operator Layer| quantizations, IO and other basic components                                                 | PQ,SQ,MemoryIO,SSDIO,......|

this abstract layered approach may facilitate scalability and code reuse
currently we only have index layer, so the issue will provide Datacell Layer and Operator Layer in this feature.
the related pull requests:

- #25 
- #34 
- #37 
- #41 
- #35 
- #59 
- #60 
- #61 
- #62 
- #79 
- #80 
- #96 
- #111 
- #114 
- #125 
todo tasks:
- support mixture DataCell
- adapt Graph DataCell for HNSW
- implement HNSW + SQ4 + SQ8 Index

### PR DESCRIPTION
**Pull request by LHT129**
#40 
- add basic test for hgraph
- trans allocator* to shard_ptr for all
- add range_search for hgraph
- rename some ut description
- fix typo of HGraph

### CODE REVIEW

File: src/index/hgraph_index.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ HGraphIndex::Init() {
     mult_ = 1 / log(1.0 * static_cast<double>(this->bottom_graph_->MaximumDegree()));
 
     this->pool_ = std::make_shared<hnswlib::VisitedListPool>(
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:81]: be careful the `VisitedListPool` object and `allocator_` object destroy sequence
here can let VisitedListPool to hold shared_ptr ?

File: src/index/hgraph_index.cpp [hunk 2]
```diff
@@ -117,42 +122,128 @@ HGraphIndex::knn_search(const DatasetPtr& query,
...
+            fmt::format("query.dim({}) must be equal to index.dim({})", query_dim, dim_));
+        // check k
+        CHECK_ARGUMENT(k > 0, fmt::format("k({}) must be greater than 0", k));
+        k = std::min(k, GetNumElements());
+
+        // check query vector
+        CHECK_ARGUMENT(query->GetNumElements() == 1, "query dataset should contain 1 vector only");
+        InnerSearchParam search_param;
+        search_param.ep_ = this->entry_point_id_;
+        search_param.ef_ = 1;
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:138]: Q: why set ef_ = 1 ?

File: src/index/hgraph_index.cpp [hunk 3]
```diff
@@ -117,42 +122,128 @@ HGraphIndex::knn_search(const DatasetPtr& query,
...
+        auto search_result = this->search_one_graph(query->GetFloat32Vectors(),
+                                                    this->bottom_graph_,
+                                                    this->basic_flatten_codes_,
+                                                    search_param);
 
-    auto ef = params.ef_search;
-    auto search_result = this->search_one_graph(
-        query->GetFloat32Vectors(), this->bottom_graph_, this->basic_flatten_codes_, ep, ef, &ft);
+        while (search_result.size() > k) {
+            search_result.pop();
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:157]: here we can add search_result.trim(k) operation ?

File: src/index/hgraph_index.cpp [hunk 4]
```diff
@@ -117,42 +122,128 @@ HGraphIndex::knn_search(const DatasetPtr& query,
...
+
+        auto* ids = (int64_t*)allocator_->Allocate(sizeof(int64_t) * search_result.size());
+        dataset_results->Ids(ids);
+        auto* dists = (float*)allocator_->Allocate(sizeof(float) * search_result.size());
+        dataset_results->Distances(dists);
+
+        for (int64_t j = search_result.size() - 1; j >= 0; --j) {
+            dists[j] = search_result.top().first;
+            ids[j] = search_result.top().second;
+            search_result.pop();
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:239]: MaxHeap not support index access ?

File: src/index/hgraph_index.h [hunk 5]
```diff
@@ -109,7 +110,8 @@ class HGraphIndex : public Index {
                 const std::string& parameters,
                 const std::function<bool(int64_t)>& filter,
```
Comment (jiaweizone) [inline src/index/hgraph_index.h:108]: filter -> callback

File: src/index/hgraph_index.h [hunk 6]
```diff
@@ -109,7 +110,8 @@ class HGraphIndex : public Index {
...
                 int64_t limited_size) const override {
-        return Dataset::Make();
+        BitsetOrCallbackFilter ft(filter);
```
Comment (jiaweizone) [inline src/index/hgraph_index.h:110]: filter(callback) ?

File: src/index/hgraph_zparameters.cpp [hunk 7]
```diff
@@ -48,6 +56,16 @@ HGraphParameters::ParseStringParam(const std::string& str) {
 void
 HGraphParameters::CheckAndSetKeyValue(const std::string& key, nlohmann::json& value) {
     const auto& iter = EXTERNAL_MAPPING.find(key);
+
+    if (key == HGRAPH_BASE_QUANTIZATION_TYPE) {
+        std::string value_str = value;
+        if (value_str != "sq8" && value_str != "fp32") {
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.cpp:63]: define SQ8 / FP32 constants ?

File: src/index/hgraph_zparameters.cpp [hunk 8]
```diff
@@ -56,7 +74,7 @@ HGraphParameters::CheckAndSetKeyValue(const std::string& key, nlohmann::json& va
         }
         *json = value;
     } else {
-        // TODO(LHT): Error logger
+        throw std::invalid_argument(fmt::format("HGraph have no config param: {}", key));
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.cpp:78]: only print warn log ?

File: src/index/hgraph_zparameters.cpp [hunk 9]
```diff
@@ -72,7 +90,7 @@ const std::string HGraphParameters::DEFAULT_HGRAPH_PARAMS = format_map(
             "type": "nsw",
             "{GRAPH_PARAMS_KEY}": {
                 "{GRAPH_PARAM_MAX_DEGREE}": 64,
-                "{GRAPH_PARAM_INIT_MAX_CAPACITY}": 20000000
+                "{GRAPH_PARAM_INIT_MAX_CAPACITY}": 10000
```
Comment (jiaweizone) [inline src/index/hgraph_zparameters.cpp]: const string  define move to file head ?

File: src/index/index_common_param.cpp [hunk 10]
```diff
@@ -78,6 +78,8 @@ IndexCommonParam::CheckAndCreate(const std::string& json_string) {
                    fmt::format("parameters[{}] must be integer type", PARAMETER_DIM));
     int64_t dim = params[PARAMETER_DIM];
     CHECK_ARGUMENT(dim > 0, fmt::format("parameters[{}] must be greater than 0", PARAMETER_DIM));
+    CHECK_ARGUMENT(dim < 65536,
```
Comment (jiaweizone) [inline src/index/index_common_param.cpp:81]: 65536 is our hard limit ?

File: src/index/hgraph_index.cpp (hunk 2 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:138]: for high level graph search

File: src/index/hgraph_index.cpp (hunk 1 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:81]: I think you are right, I want remove the dependance of the hnswlib in hgraph, another pr #123 will introduce new visited pool.  now fix it on hnswlib's visited_pool

Comment (jiaweizone) [reply src/index/hgraph_index.cpp:81]: OK

**Pull request closed without merging**